scapy/contrib/automotive/gm/gmlan.py

```diff
@@ -0,0 +1,644 @@
+    fields_desc = [
+        XByteEnumField('service', 0, services)
+    ]
+
+    def answers(self, other):
+        """DEV: true if self is an answer from other"""
+        if other.__class__ == self.__class__:
+            return (other.service + 0x40) == self.service or \
+                   (self.service == 0x7f and
+                    (self.requestServiceId == other.service))
+        return 0
+
+    def hashret(self):
+        if self.service == 0x7f:
+            return struct.pack('B', self.requestServiceId)
+        return struct.pack('B', self.service & ~0x40)
+
+
+# ########################IDO###################################
+class GMLAN_IDO(Packet):
+    subfunctions = {
+        0x02: 'disableAllDTCs',
+        0x03: 'enableDTCsDuringDevCntrl',
+        0x04: 'wakeUpLinks'}
+    name = 'InitiateDiagnosticOperation'
+    fields_desc = [
+        ByteEnumField('subfunction', 0, subfunctions)
+    ]
+
+
+bind_layers(GMLAN, GMLAN_IDO, service=0x10)
+
+
+# ########################RFRD###################################
+class GMLAN_DTC(Packet):
+    name = 'GMLAN DTC information'
+    fields_desc = [
+        XByteField('failureRecordNumber', 0),
+        XByteField('DTCHighByte', 0),
+        XByteField('DTCLowByte', 0),
+        XByteField('DTCFailureType', 0)
+    ]
+
+    def extract_padding(self, p):
+        return "", p
+
+
+class GMLAN_PID(Packet):
+    name = 'GMLAN PID information'
+    fields_desc = [
+        XByteField('PIDHB', 0),
+        XByteField('PIDLB', 0)
+    ]
+
+    def extract_padding(self, p):
+        return "", p
+
+
+class GMLAN_RFRD(Packet):
+    subfunctions = {
+        0x01: 'readFailureRecordIdentifiers',
+        0x02: 'readFailureRecordParameters'}
+    name = 'ReadFailureRecordData'
+    fields_desc = [
+        ByteEnumField('subfunction', 0, subfunctions),
+        ConditionalField(PacketField("dtc", b'', GMLAN_DTC),
+                         lambda pkt: pkt.subfunction == 0x02)
+    ]
+
+
+bind_layers(GMLAN, GMLAN_RFRD, service=0x12)
+
+
+class GMLAN_RFRDPR(Packet):
+    name = 'ReadFailureRecordDataPositiveResponse'
+    fields_desc = [
+        ByteEnumField('subfunction', 0, GMLAN_RFRD.subfunctions)
+    ]
+
+
+bind_layers(GMLAN, GMLAN_RFRDPR, service=0x52)
+
+
+class GMLAN_RFRDPR_RFRI(Packet):
+    failureRecordDataStructureIdentifiers = {
+        0x00: "PID",
+        0x01: "DPID"
+    }
+    name = 'ReadFailureRecordDataPositiveResponse_readFailureRecordIdentifiers'
+    fields_desc = [
+        ByteEnumField('failureRecordDataStructureIdentifier', 0,
+                      failureRecordDataStructureIdentifiers),
+        PacketListField("dtcs", [], GMLAN_DTC)
+    ]
+
+
+bind_layers(GMLAN_RFRDPR, GMLAN_RFRDPR_RFRI, subfunction=0x01)
+
+
+class GMLAN_RFRDPR_RFRP(Packet):
+    name = 'ReadFailureRecordDataPositiveResponse_readFailureRecordParameters'
+    fields_desc = [
+        PacketField("dtc", b'', GMLAN_DTC),
+        PacketListField("pids", [], GMLAN_PID)
+    ]
```

Comment:
How about
```python

class GMLAN_DPID(Packet):
    name = 'GMLAN PID information'
    fields_desc = [
        [... magic stuff goes here...]
    ]

    def extract_padding(self, p):
        return "", p


class GMLAN_RFRDPR_RFRP(Packet):
    name = 'ReadFailureRecordDataPositiveResponse_readFailureRecordParameters'
    fields_desc = [
        PacketField("dtc", b'', GMLAN_DTC),
        MultipleTypeField([
            (PacketListField("pids", [], GMLAN_PID), lambda pkt: pkt.underlayer and pkt.underlayer.failureRecordDataStructureIdentifier == 0),
            (PacketListField("pids", [], GMLAN_DPID), lambda pkt: pkt.underlayer and pkt.underlayer.failureRecordDataStructureIdentifier == 1),
        ], PacketListField("pids", [], GMLAN_PID))
    ]
```

Reply:
Hi, thanks for your help here. Unfortunately, the decision between PID and DPID can not be done by the data in the packet. You have to know, what the ECU supports, in order to decide if the data has to be interpreted as PID or DPID. I just realised that by double checking some documentation. So I will make a configuration for that decision.

Reply:
After reading the documentation again, even GMLAN_PID is wrong. PID and DPID can have a length between 1 and 7 bytes, and the definition is based on a specific ECU. So I will leave that part as padding. One can add a specific description based on the PID or the DTC, later.